Optimize Triton template heuristics

robieta opened the pull request:
Summary:
This diff contains three small optimizations:
1) Directly cache the triton Config object import. Not a huge win, but measurably faster than relying on importlib's cache.

2) Only copy configs when the new value is different from the old one. Configs are fairly large objects, so unneccesary dict copies get expensive.

3) Replace `gcd(k, BLOCK_K) == BLOCK_K` with `(k % BLOCK_K) == 0`. This is equivalent when `BLOCK_K > 0`, which must be true.

Test Plan:
```
tlp buck run mode/opt //scripts/paulzhan:repro
```
and then looking at perfetto.

Differential Revision: D88415189




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

robieta commented:
@pytorchbot rebase

shunting314 commented on torch/_inductor/template_heuristics/triton.py:
```diff
@@ -211,6 +211,14 @@ def __call__(
+    """Import has a cache, but this is faster."""
+    from triton import Config as TritonConfig
+
+    return TritonConfig
```
Why not: `from torch._inductor.runtime.triton_compat import Config` ?

robieta committed:
Optimize Triton template heuristics (#170444)

Summary:

This diff contains three small optimizations:
1) Directly cache the triton Config object import. Not a huge win, but measurably faster than relying on importlib's cache.

2) Only copy configs when the new value is different from the old one. Configs are fairly large objects, so unneccesary dict copies get expensive.

3) Replace `gcd(k, BLOCK_K) == BLOCK_K` with `(k % BLOCK_K) == 0`. This is equivalent when `BLOCK_K > 0`, which must be true.

Test Plan:
```
tlp buck run mode/opt //scripts/paulzhan:repro
```
and then looking at perfetto.

Reviewed By: eellison, PaulZhang12, shunting314

Differential Revision: D88415189

robieta commented:
@pytorchbot merge
